### PR TITLE
Server side rendering

### LINKED ISSUE #406: Migrate React-router to server side

### PR DESCRIPTION
**Pull request by alanmoo**
Closes #406

### CODE REVIEW

**Comment by gvn** [comment]
I think we should get two reviewers on this due to the size and complexity.

@Pomax, you have a lot of experience w. server side React rendering, would you mind giving this a review as well?

File: app/index.jsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import React from "react";
+
+module.exports = React.createClass({
+  render(){
+    return(
+<html>
```
Comment (gvn) [inline app/index.jsx]: Can we properly indent this block?

**Comment by gvn** [comment]
The webpack config files have some crazy indentation going on. Can you fix that?

**Comment by gvn** [comment]
When I load [http://127.0.0.1:1818/programs/studygroups](http://127.0.0.1:1818/programs/studygroups) I get this error on the server:

```
{ Error: Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.
    at a (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:7329)
    at u (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:22476)
    at d.Mixin.getNativeNode (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:16:21568)
    at Object.s.getNativeNode (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:30054)
    at d.v.Mixin._updateChildren (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:17:18561)
    at d.v.Mixin.updateChildren (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:17:18166)
    at d.Mixin._updateDOMChildren (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:16:21513)
    at d.Mixin.updateComponent (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:16:19904)
    at d.Mixin.receiveComponent (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:16:19331)
    at Object.s.receiveComponent (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:30297) framesToPop: 1 }
{ Error: Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.
    at a (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:7329)
    at Object.u [as getNodeFromInstance] (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:22476)
    at a.getNativeNode (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:18:3915)
    at Object.s.getNativeNode (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:30054)
    at d.v.Mixin._updateChildren (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:17:18561)
    at d.v.Mixin.updateChildren (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:17:18166)
    at d.Mixin._updateDOMChildren (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:16:21513)
    at d.Mixin.updateComponent (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:16:19904)
    at d.Mixin.receiveComponent (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:16:19331)
    at Object.s.receiveComponent (/Volumes/Nostromo/mozilla-code/science.mozilla.org/server.js:1:30297) framesToPop: 1 }
```

**Comment by alanmoo** [comment]
Investigating the error. After un=minifying, it seems to be related to https://github.com/facebook/react/issues/6232

<details>
<summary> Invariant Violation</summary>
{ [Invariant Violation: React DOM tree root should always have a node reference.] name: 'Invariant Violation', framesToPop: 1 }
</details>

**Comment by alanmoo** [comment]
Unless I broke something with the deploy process, this should be ready to review again.

**Comment by alanmoo** [comment]
@alicoding this doesn't deploy properly on heroku for some reason, can you take a look?

**Comment by alicoding** [comment]
Is it working now?

**Comment by gvn** [comment]
I'm seeing: 

screen shot 2016-07-08 at 9 53 53 am

File: webpack.server.config.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  externals: nodeModules,
+  output: {
+    filename: `/server.js`,
+    path: `${__dirname}`
+  },
+  plugins: [
+    new webpack.DefinePlugin({
+      'process.env': {
+        'NODE_ENV': JSON.stringify(`production`)
```
Comment (alicoding) [inline webpack.server.config.js]: Also add this `'PORT': process.env.PORT`

File: app/server.jsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import React from 'react';
+
+import ReactDOMServer from 'react-dom/server';
+import { match, RouterContext } from 'react-router';
+import routes from './react/routes.jsx';
+import HTML from './index.jsx';
+
+
+app.listen(process.env.PORT || 1818, function () {
+  console.log(`Listening on port 1818!`);
```
Comment (alicoding) [inline app/server.jsx]: Don't hardcode that port in the comment.

**Comment by alanmoo** [comment]
@alicoding I made those changes but it's still failing.

File: app/server.jsx [hunk 4]
```diff
@@ -11,7 +11,7 @@ import HTML from './index.jsx';
 
 
 app.listen(process.env.PORT || 1818, function () {
```
Comment (alicoding) [inline app/server.jsx]: Change this whole block to this:

```
var listener = app.listen(process.env.PORT || 1818, function(){
    console.log(`Listening on port ${listener.address().port}!`);
});
```

**Comment by alicoding** [comment]
The problem here is that `process.env.PORT` is not there when `server.js` is trying to read it and it's being default to `1818` which is not going to work with heroku since they're using their own port number here...

```
$ heroku run bash -a science-mozilla-org-sta-pr-512
Running bash on ⬢ science-mozilla-org-sta-pr-512... up, run.4421
~ $ echo $PORT
32647
```

**Comment by alanmoo** [comment]
After removing the server file optimizations (since that file's size doesn't really matter), it looks like webpack is trying to do too much voodoo on the file, causing it to optimize out env vars (at least locally). Investigating what I can do to avoid this now.

**Comment by alanmoo** [comment]
This is now rendering properly, but when landing on https://science-mozilla-org-sta-pr-512.herokuapp.com/projects, a 500 error results. This seems to come from the Debounce component (https://github.com/alanmoo/science.mozilla.org/blob/optimize-webpack/app/react/pages/projects/projects.jsx#L125), it's probably attempting to add an event listener server side. I'm not quite sure how to solve this though. Anyone have ideas?

**Comment by alicoding** [comment]
This is the error from heroku logs

```
2016-07-12T13:21:53.598192+00:00 app[web.1]:     at basePickBy (/app/node_modules/lodash/lodash.js:3347:13)
2016-07-12T13:21:53.598194+00:00 app[web.1]:     at Function.pickBy (/app/node_modules/lodash/lodash.js:12917:36)
2016-07-12T13:21:53.598144+00:00 app[web.1]:     at Base._shouldWrapHandler (/app/node_modules/react-throttle/lib/classes/processors/Base.js:51:78)
2016-07-12T13:21:53.598128+00:00 app[web.1]: TypeError: _this.handlersToWrap.includes is not a function
2016-07-12T13:21:53.598194+00:00 app[web.1]:     at Debouncer.Base._extractHandlers (/app/node_modules/react-throttle/lib/classes/processors/Base.js:43:16)
2016-07-12T13:21:53.598195+00:00 app[web.1]:     at Debouncer.Base._wrapHandlers (/app/node_modules/react-throttle/lib/classes/processors/Base.js:47:32)
2016-07-12T13:21:53.598196+00:00 app[web.1]:     at Debouncer.Base.element (/app/node_modules/react-throttle/lib/classes/processors/Base.js:34:46)
2016-07-12T13:21:53.598197+00:00 app[web.1]:     at ReactCompositeComponentWrapper._renderValidatedComponentWithoutOwnerOrContext (/app/server.js:12447:35)
2016-07-12T13:21:53.598195+00:00 app[web.1]:     at Debouncer.Base._processElement (/app/node_modules/react-throttle/lib/classes/processors/Base.js:38:30)
2016-07-12T13:21:53.598196+00:00 app[web.1]:     at Debounce.Base._this.render (/app/node_modules/react-throttle/lib/components/Base.js:30:31)
2016-07-12T13:21:53.598197+00:00 app[web.1]:     at ReactCompositeComponentWrapper._renderValidatedComponent (/app/server.js:12467:33)
2016-07-12T13:21:53.955621+00:00 app[web.1]: { [Error: Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.] framesToPop: 1 }
```

**Comment by alanmoo** [comment]
Yeah, I spotted that- it's the debounce component having issues server-side. Investigating how to solve for this other than removing the component or forking it to include a check for its environment.

**Comment by alanmoo** [comment]
I believe I've got this working now. Check out the latest deployment

**Comment by alanmoo** [comment]
Squashed these commits into one based on master, and added a commit to bring in the Google Analytics from #560

**Pull request closed without merging**